src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/events/EnvelopeHandler.java

```diff
@@ -35,12 +35,15 @@ public void handleEnvelope(Envelope envelope) {
 
                 if (caseDetailsFound.isPresent()) {
                     CaseDetails existingCase = caseDetailsFound.get();
-                    evidenceAttacher.attach(envelope, existingCase);
-                    paymentsProcessor.createPayments(envelope, existingCase.getId(), false);
+                    boolean docsAttached = evidenceAttacher.attach(envelope, existingCase);
+                    if (docsAttached) {
+                        paymentsProcessor.createPayments(envelope, existingCase.getId(), false);
```

Comment:
Should payment creation be unconditional, i.e. happen in both cases?

Reply:
If we don't attach scanned docs then we should not update payments as the caseworker would not be aware or may need documentation corresponding to payments(for e.g evidence for change of name or address)

Reply:
After creating the exception record, the payment message is published with the exception record id. Is this ok?

Reply:
Ok, there's some lack of clarity here. From the code `docsAttached` is `false` if we failed to attach them to the case, getting a 404 response. If there were no documents to be attached, the method will return `true`.

I guess we have too little information coming from this method, because:
- if nothing failed, but there were no docs attached, we shouldn't create payments
- if at least one doc was attached, we should create payments
- if we failed because a 404 from CCD, we should create an exception record, but also payments

Is this what we should expect?

Reply:
I missed the `no documents to attach` case.  I don't think we have to create an Exception record when there are no documents to attach.

Reply:
Exception handling way would be better imo way Rafal suggested

Reply:
> If we don't attach scanned docs then I think it is better we don't update payments as the caseworker would not be aware or may need documentation corresponding to payments(for e.g evidence for change of name or address)

How about the case when we previously attached supplementary case from this envelope to the case and failed to publish payments? The envelope returned to the queue and is processed again now. This time, we won't attach any documents to the case, as it has already happened, but we still haven't published payments. Given payment creation is idempotent in PayHub, so publishing them twice wouldn't cause any harm.

Reply:
If we have attached the docs or skipped attaching because we have already attached previously then we should process the payments. Only when we fail to attach due to access issue or configuration issue(404/403) we should create an exception record and publish payments.
Transaction T1 -> Attach scanned docs or create exception record
Transaction T2->  Process payments 

Also the way T2 would work is also impacted what you do in T1 as we need to publish different payment messages.

Reply:
So as long as we're not having a general failure resulting in not completing the envelope, we should always publish payments. Is that correct?

Reply:
yes correct. Also we will have to keep in mind idempotency scenario for exception record creation(probably we can handle it as a separate PR else it will make this PR complicated)

Reply:
Actually, the code already works as expected.

Reply:
> yes correct. Also we will have to keep in mind idempotency scenario for exception record creation(probably we can handle it as a separate PR else it will make this PR complicated)

@nitinprabhu As discussed earlier, we are already checking if an Exception record was created for an envelope
https://github.com/hmcts/bulk-scan-orchestrator/blob/28d55bfd9c583dc96a71f0048c07359e60f073f6/src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/events/CreateExceptionRecord.java#L48